core/modules/parser/MySqlListener.cc

```diff
@@ -0,0 +1,1328 @@
+};
+
+
+class SelectColumnElementAdapter : public Adapter, public FullColumnNameCBH {
+public:
+    SelectColumnElementAdapter(shared_ptr<SelectColumnElementCBH> parent, antlr4::ParserRuleContext* ctx)
+    : Adapter(parent, ctx) {}
+
+    void handleFullColumnName(shared_ptr<query::ValueExpr> columnValueExpr) override {
+        lockedParent<SelectColumnElementCBH>()->handleColumnElement(columnValueExpr);
+    }
+};
+
+
+class UidAdapter : public Adapter {
+public:
+    UidAdapter(shared_ptr<UidCBH> parent, MySqlParser::UidContext* ctx)
+    : Adapter(parent, ctx), _uidContext(ctx) {}
+
+    void onExit() override {
+        LOGS(_log, LOG_LVL_DEBUG, __FUNCTION__);
+        // Fetching the string from a Uid shortcuts a large part of the syntax tree defined under Uid
+        // (see MySqlParser.g4). If Adapters for any nodes in the tree below Uid are implemented then
+        // it will have to be handled and this shortcut may not be taken.
+        lockedParent<UidCBH>()->handleUidString(_uidContext->getText());
+    }
+
+private:
+    MySqlParser::UidContext* _uidContext;
+};
+
+
+class ConstantAdapter : public Adapter {
+public:
+    ConstantAdapter(shared_ptr<ConstantCBH> parent, MySqlParser::ConstantContext* ctx)
+    : Adapter(parent, ctx), _constantContext(ctx) {}
+
+    void onExit() override {
+        lockedParent<ConstantCBH>()->handleConstant(_constantContext->getText());
+    }
+
+private:
+    MySqlParser::ConstantContext* _constantContext;
+};
+
+
+/// MySqlListener impl
+
+
+MySqlListener::MySqlListener() {
+    _rootAdapter = std::make_shared<RootAdapter>();
+    _adapterStack.push(_rootAdapter);
+}
+
+
+std::shared_ptr<query::SelectStmt> MySqlListener::getSelectStatement() const {
+    return _rootAdapter->getSelectStatement();
+}
+
+
+// Create and push an Adapter onto the context stack, using the current top of the stack as a callback handler
+// for the new Adapter. Returns the new Adapter.
+template<typename ParentCBH, typename ChildAdapter, typename Context>
+std::shared_ptr<ChildAdapter> MySqlListener::pushAdapterStack(Context * ctx) {
+    auto p = std::dynamic_pointer_cast<ParentCBH>(_adapterStack.top());
+    if (nullptr == p) {
+        int status;
+        std::ostringstream msg;
+        msg << "can't acquire expected Adapter " <<
+                abi::__cxa_demangle(typeid(ParentCBH).name(),0,0,&status) <<
+                " from top of listenerStack.";
+        LOGS(_log, LOG_LVL_ERROR, msg.str());
+        throw adapter_order_error(msg.str());
+    }
+    auto childAdapter = std::make_shared<ChildAdapter>(p, ctx);
+    _adapterStack.push(childAdapter);
+    childAdapter->onEnter();
+    return childAdapter;
+}
+
+
+template<typename ChildAdapter>
+void MySqlListener::popAdapterStack() {
+    shared_ptr<Adapter> adapterPtr = _adapterStack.top();
+    adapterPtr->onExit();
+    _adapterStack.pop();
+    shared_ptr<ChildAdapter> derivedPtr = dynamic_pointer_cast<ChildAdapter>(adapterPtr);
```

Comment:
Do you really need to check the type here, dynamic casts are not free?

Reply:
it's not required, I put it in as a sanity check, in case the enter/exit funcs get called out of order. IMO it's nice to have at least for now.

Reply:
I'll add a comment.